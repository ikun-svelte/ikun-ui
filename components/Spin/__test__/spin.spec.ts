import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { tick } from 'svelte';
import KSpinShow from './fixture/spin.show.svelte';
import KSpinText from './fixture/spin.text.svelte';
import KSpinFull from './fixture/spin.full.svelte';
import KSpinBg from './fixture/spin.bg.svelte';
import KSpinSpinner from './fixture/spin.spinner.svelte';
import KSpinRotate from './fixture/spin.rotate.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
	vi.useFakeTimers();
});
afterEach(() => {
	host.remove();
	vi.restoreAllMocks();
});

describe('Test: KSpin', () => {
	test('options: show', async () => {
		const instance = new KSpinShow({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-spin--spinner')).not.toBeTruthy();
		const btn = document.getElementById('spin_show');
		btn?.click();
		await tick();
		expect(host.innerHTML.includes('k-spin--spinner')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('options: text', async () => {
		const instance = new KSpinText({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-spin--spinner')).not.toBeTruthy();
		expect(host.innerHTML.includes('多看一眼就会爆炸！')).not.toBeTruthy();
		const btn = document.getElementById('spin_text');
		btn?.click();
		await tick();
		expect(host.innerHTML.includes('k-spin--spinner')).toBeTruthy();
		expect(host.innerHTML.includes('多看一眼就会爆炸！')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('options: fullScreen', async () => {
		const instance = new KSpinFull({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		expect(document.body.innerHTML.includes('k-spin--spinner')).not.toBeTruthy();
		expect(document.body.innerHTML.includes('k-mask--base')).not.toBeTruthy();
		const btn = document.getElementById('spin_full');
		btn?.click();
		await tick();
		expect(document.body.innerHTML.includes('k-spin--spinner')).toBeTruthy();
		expect(document.body.innerHTML.includes('k-mask--base')).toBeTruthy();
		await vi.advanceTimersByTimeAsync(320);
		expect(
			document.body.innerHTML.includes(
				'top: 0px; left: 0px; width: 100%; height: 100%; position: fixed;'
			)
		).toBeTruthy();
		expect(document.body).matchSnapshot();
	});

	test('options: background', async () => {
		const instance = new KSpinBg({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-spin--spinner')).not.toBeTruthy();
		const btn = document.getElementById('spin_bg');
		btn?.click();
		await tick();
		expect(host.innerHTML.includes('k-spin--spinner')).toBeTruthy();
		await vi.advanceTimersByTimeAsync(320);
		expect(host.innerHTML.includes('background-color: red;')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('options: spinner', async () => {
		const instance = new KSpinSpinner({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-spin--spinner')).not.toBeTruthy();
		const btn = document.getElementById('spin_spinner');
		btn?.click();
		await tick();
		expect(host.innerHTML.includes('k-spin--spinner')).toBeTruthy();
		expect(host.innerHTML.includes('custom spinner')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('options: rotating', async () => {
		const instance = new KSpinRotate({
			target: host
		});
		await tick();
		expect(instance).toBeTruthy();
		expect(host.innerHTML.includes('k-spin--spinner')).not.toBeTruthy();
		const btn = document.getElementById('spin_rotate');
		btn?.click();
		await tick();
		expect(host.innerHTML.includes('k-spin--spinner')).toBeTruthy();
		expect(host.innerHTML.includes('k-spin--spinner__rotating')).not.toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});
});
