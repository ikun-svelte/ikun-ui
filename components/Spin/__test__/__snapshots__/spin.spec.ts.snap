// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KSpin > options: background 1`] = `"<div class="w-full h-200px flex justify-center items-center" style="overflow: hidden; position: relative;"><button id="spin_bg">Spin</button><div role="presentation" style="background-color: red; top: 0px; left: 0px; width: 100%; height: 100%;" class="k-mask--base"><div class=""><div class="k-spin--spinner"><div class="k-spin--spinner__rotating"><span class="k-icon--base k-icon-transition i-carbon-circle-dash text-ikun-main" role="" style="width: 24px; height: 24px;"></span></div> <p class="k-spin--text">loading...</p></div></div></div></div>"`;

exports[`Test: KSpin > options: fullScreen 1`] = `
<body
  style="overflow: hidden; position: relative;"
>
  <div
    id="host"
  >
    <div
      class="w-full h-200px flex justify-center items-center"
    >
      <button
        id="spin_full"
      >
        Spin
      </button>
    </div>
  </div>
  <div
    class="k-mask--base"
    role="presentation"
    style="top: 0px; left: 0px; width: 100%; height: 100%; position: fixed;"
  >
    <div
      class=""
    >
      <div
        class="k-spin--spinner"
      >
        <div
          class="k-spin--spinner__rotating"
        >
          <span
            class="k-icon--base k-icon-transition i-carbon-circle-dash text-ikun-main"
            role=""
            style="width: 24px; height: 24px;"
          />
        </div>
         
        <p
          class="k-spin--text"
        >
          loading...
        </p>
      </div>
    </div>
  </div>
  
  
</body>
`;

exports[`Test: KSpin > options: rotating 1`] = `"<div class="w-full h-200px flex justify-center items-center"><button id="spin_rotate">Spin</button><div role="presentation" class="k-mask--base" style="top: 0px; left: 0px; width: 100%; height: 100%;"><div class=""><div class="k-spin--spinner"><div><span class="k-icon--base k-icon-transition i-carbon-circle-dash text-ikun-main" role="" style="width: 24px; height: 24px;"></span></div> <p class="k-spin--text">loading...</p></div></div></div></div>"`;

exports[`Test: KSpin > options: show 1`] = `"<div class="w-full h-200px flex justify-center items-center"><button id="spin_show">Spin</button><div role="presentation" class="k-mask--base" style="top: 0px; left: 0px; width: 100%; height: 100%;"><div class=""><div class="k-spin--spinner"><div class="k-spin--spinner__rotating"><span class="k-icon--base k-icon-transition i-carbon-circle-dash text-ikun-main" role="" style="width: 24px; height: 24px;"></span></div> <p class="k-spin--text">loading...</p></div></div></div></div>"`;

exports[`Test: KSpin > options: spinner 1`] = `"<div class="w-full h-200px flex justify-center items-center"><button id="spin_spinner">Spin</button><div role="presentation" class="k-mask--base" style="top: 0px; left: 0px; width: 100%; height: 100%;"><div class=""><div class="k-spin--spinner"><div class="k-spin--spinner__rotating"><span class="k-icon--base k-icon-transition custom spinner text-ikun-main" role="" style="width: 24px; height: 24px;"></span></div> <p class="k-spin--text">loading...</p></div></div></div></div>"`;

exports[`Test: KSpin > options: text 1`] = `"<div class="w-full h-200px flex justify-center items-center"><button id="spin_text">Spin</button><div role="presentation" class="k-mask--base" style="top: 0px; left: 0px; width: 100%; height: 100%;"><div class=""><div class="k-spin--spinner"><div class="k-spin--spinner__rotating"><span class="k-icon--base k-icon-transition i-carbon-circle-dash text-ikun-main" role="" style="width: 24px; height: 24px;"></span></div> <p class="k-spin--text">多看一眼就会爆炸！</p></div></div></div></div>"`;
